File: Data Structures & Algorithms/subsets/submission-2.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        def rec(i, l):
            res.append(l[:])
            for j in range(i, len(nums)):
                n = nums[j]
                l.append(n)
                rec(j+1, l)
                l.pop()
        rec(0, [])
        return res

File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> numToFreq = new HashMap<>();
        for (int n : nums) {
            numToFreq.put(n, numToFreq.getOrDefault(n, 0) + 1);
        }
        List<List<Integer>> freqToNum = new ArrayList<>(nums.length + 1);
        for (int i = 0; i < nums.length + 1; i++) {
            freqToNum.add(new ArrayList<>());
        }
        for (Map.Entry<Integer, Integer> entry : numToFreq.entrySet()) {
            freqToNum.get(entry.getValue()).add(entry.getKey());
        }
        int[] res = new int[k];
        int res_i = 0;
        for (int i = nums.length; i >= 0; i--) {
            List<Integer> bucket = freqToNum.get(i);
            for (int n : bucket) {
                res[res_i] = n;
                res_i++;
                if (res_i == k) {
                    return res;
                }
            }
        }
        return res;
    }
}
